### PR TITLE
M3-160 Search Landing

### PR DESCRIPTION
**Pull request by martinmckenna**
### Purpose

Direct users to Search Landing page upon hitting the enter key in the search bar

### Notes

* Unlike the wireframes, each entity is split into separate panels.
* User can only access this page after typing 3 or more characters into the search bar and hitting enter. Typing anything less will do nothing
* Upon mount, `<SearchLanding />` makes a `promise.all` request. If one of these fails, the entire page will error out

### Known Issues

- [ ] Search Result row missing additional information. Only showing label of the entity, at the moment

### CODE REVIEW

**Comment by asauber** [comment]
We have icons for StackScripts and Domains now. `stackscripts.svg` and `domain.svg`

**Comment by Jskobos** [comment]
Addressed the first issue above (searching when already on the search landing page). The SS and Domain icons are present, but seem to be blank gray circles.

Refactoring etc. probably necessary.

**Comment by martinmckenna** [comment]
blocked pending #3604

File: src/features/Search/SearchLanding.tsx [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+    const { query: { query } } = this.state;
+
+    this.setState({ isLoading: true });
+
+    Promise.all([
+      getLinodes(
+        { page: 1, page_size: 25 },
+        getFilter('linode', query)
+      ),
+      getVolumes(
```
Comment (asauber) [inline src/features/Search/SearchLanding.tsx]: getVolumes does not properly `setXFilter` so this always returns all volumes

File: src/features/Search/SearchLanding.tsx [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+    case 'volume':
+    case 'nodebalancer':
+      return {
+        "label": {
+          ["+contains"]: filterTerm
+        },
+      }
+    case 'stackscript':
+      return {
+        ["+or"]: [{
```
Comment (asauber) [inline src/features/Search/SearchLanding.tsx]: I realized that the brackets around these `"+"` keys are not necessary. For example, the following is a valid JavaScript object, even in ES5.

`{"+or": [{"label": { "+contains": "hello" }}, {"label": { "+contains": "world" }}]}`

File: src/features/Search/SearchLanding.tsx [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+  /*
+  * START get more update handlers
+  * 
+  * Basically, when a user changes the page size or requests the next
+  * page of search results, we need to both update the displayed data
+  * and update the actual number of results displayed, so that we can
+  * display to the user, "Showing 23 out of 100 results"
+  */
+
+  handleGetMoreLinodes = (newData: Linode.ResourcePage<Linode.Linode>) => {
```
Comment (asauber) [inline src/features/Search/SearchLanding.tsx]: These functions could be abstracted into a single function that takes a `resultType` parameter.

File: src/features/TopMenu/SearchBar/SearchBar.tsx [hunk 4]
```diff
@@ -337,6 +337,22 @@ class SearchBar extends React.Component<FinalProps, State> {
...
+    const { searchText } = this.state;
+    /*
+    * If we hit the enter key and something
+    * has been typed into the search bar
+    */
+    if (e.keyCode === 13 && searchText && searchText.length >= 3) {
+      const location = history.location.pathname;
+      history.push(`/search?query=${searchText}`);
+      if (location === '/search') {
+        history.go(0);
```
Comment (asauber) [inline src/features/TopMenu/SearchBar/SearchBar.tsx]: Using `go` seems to reload the whole app. Instead we can wrap `SearchLanding` in the `reloadableWithRouter` HOC.

**Comment by rjstires** [comment]
I know this is not the time to discuss UX, but I wonder if it's worthwhile to just use quarters of the page per entity?

**Comment by asauber** [comment]
A lot of this is better than the implementation of SearchBar itself. It uses the API to filter rather than prefetching everything. However, the UX needs work as discussed.

**Comment by martinmckenna** [comment]
closing as this feature may not see the light of day

**Pull request closed without merging**